graphs/greedy_best_first.py

```diff
@@ -136,22 +153,21 @@ def get_successors(self, parent: Node) -> list[Node]:
             pos_x = parent.pos_x + action[1]
```

Comment:
You could get the successors using a list comprehension:
```py
successors = [
    Node(
        (pos_x := parent.pos_x + action[1]),
        (pos_y := parent.pos_y + action[0]),
        self.target.pos_x,
        self.target.pos_y,
        parent.g_cost + 1,
        parent,
    )
    for action in delta
    if (
        0 <= pos_x < len(self.grid[0])
        and 0 <= pos_y < len(self.grid)
        and self.grid[pos_y][pos_x] == 0
    )
]
```

but ultimately I could go either way since this isn't necessarily more readable than what you already have.

Reply:
@tianyizheng02 
i tried but i got this failed.

```
ruff.....................................................................Failed
- hook id: ruff
- exit code: 1

graphs/greedy_best_first.py:153:18: F841 Local variable `pos_x` is assigned to but never used
    |
151 |         return [
152 |             Node(
153 |                 (pos_x := parent.pos_x + action[1]),
    |                  ^^^^^ F841
154 |                 (pos_y := parent.pos_y + action[0]),
155 |                 self.target.pos_x,
    |
    = help: Remove assignment to unused variable `pos_x`

graphs/greedy_best_first.py:154:18: F841 Local variable `pos_y` is assigned to but never used
    |
152 |             Node(
153 |                 (pos_x := parent.pos_x + action[1]),
154 |                 (pos_y := parent.pos_y + action[0]),
    |                  ^^^^^ F841
155 |                 self.target.pos_x,
156 |                 self.target.pos_y,
    |
    = help: Remove assignment to unused variable `pos_y`

Found 2 errors.

black....................................................................Passed
codespell................................................................Passed
pyproject-fmt........................................(no files to check)Skipped
Validate filenames.......................................................Passed
Validate pyproject.toml..............................(no files to check)Skipped
mypy.....................................................................Failed
- hook id: mypy
- exit code: 1

graphs/greedy_best_first.py:162: error: Name "pos_x" is used before definition  [used-before-def]
graphs/greedy_best_first.py:163: error: Name "pos_y" is used before definition  [used-before-def]
graphs/greedy_best_first.py:164: error: Name "pos_y" is used before definition  [used-before-def]
graphs/greedy_best_first.py:164: error: Name "pos_x" is used before definition  [used-before-def]
Found 4 errors in 1 file (checked 1 source file)
```

Reply:
Sorry, I think this is the correct syntax?
```py
successors = [
    Node(
        pos_x,
        pos_y,
        self.target.pos_x,
        self.target.pos_y,
        parent.g_cost + 1,
        parent,
    )
    for action in delta
    if (
        0 <= (pos_x := parent.pos_x + action[1]) < len(self.grid[0])
        and 0 <= (pos_y := parent.pos_y + action[0]) < len(self.grid)
        and self.grid[pos_y][pos_x] == 0
    )
]
```

Reply:
@tianyizheng02 
Thanks, I applied list comprehension.